libyul/optimiser/FunctionSpecializer.h

```diff
@@ -0,0 +1,104 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/optimiser/ASTWalker.h>
+
+#include <libyul/ASTForward.h>
+#include <libyul/Dialect.h>
+
+#include <optional>
+#include <vector>
+#include <map>
+
+namespace solidity::yul
+{
+
+/**
+ * FunctionSpecializer: Optimiser step that specializes the function with its literal arguments.
+ *
+ * If a function, say, `function f(a, b) { sstore (a, b)}`, is called with literal arguments, for
+ * example, `f(x, 5)`, where `x` is an identifier, it could be specialized by creating a new
+ * function `f_1` that takes only one argument, i.e.,
+ *
+ *   function f_1(a_1) {
+ *     let b_1 := 5
+ *     sstore(a_1, b_1)
+ *   }
+ *
+ * Other optimization steps will be able to make more simplifications to the function. The
+ * optimization step is mainly useful for functions that would not be inlined.
+ *
+ * Prerequisites: Disambiguator, FunctionHoister
+ *
+ * LiteralRematerialiser is recommended as a prerequisite, even though it's not required for
+ * correctness.
+ */
+class FunctionSpecializer: public ASTModifier
+{
+public:
+	/// A vector of function-call arguments. An element 'has value' if it's a literal, and the
+	/// corresponding Expression would be the literal.
+	using LiteralArguments = std::vector<std::optional<Expression>>;
+
+	static constexpr char const* name{"FunctionSpecializer"};
+	static void run(OptimiserStepContext& _context, Block& _ast);
+
+	using ASTModifier::operator();
```

Comment:
Is the using declaration needed?

Reply:
Yes. I think without it, you would get errors like these:

```
error: no match for call to ‘(solidity::yul::FunctionSpecializer) (solidity::yul::Block&)’
  126 |  f(_ast);
```